contracts/pcv/balancer/BalancerPCVDepositBase.sol

```diff
@@ -0,0 +1,168 @@
+        poolId = _poolId;
+
+        (poolAddress, ) = IVault(_vault).getPool(_poolId);
+
+        // get the balancer pool tokens
+        IERC20[] memory tokens;
+        (tokens, , ) = IVault(_vault).getPoolTokens(_poolId);
+
+        // cache the balancer pool tokens as Assets
+        poolAssets = new IAsset[](tokens.length);
+        for (uint256 i = 0; i < tokens.length; i++) {
+            poolAssets[i] = IAsset(address(tokens[i]));
+        }
+    }
+
+    // Accept ETH transfers
+    receive() external payable {}
+
+    /// @notice Wraps all ETH held by the contract to WETH
+    /// Anyone can call it.
+    /// Balancer uses WETH in its pools, and not ETH.
+    function wrapETH() external {
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance != 0) {
+            Constants.WETH.deposit{value: ethBalance}();
+        }
+    }
+
+    /// @notice unwrap WETH on the contract, for instance before
+    /// sending to another PCVDeposit that needs pure ETH.
+    /// Balancer uses WETH in its pools, and not ETH.
+    function unwrapETH() external onlyPCVController {
+        uint256 wethBalance = IERC20(address(Constants.WETH)).balanceOf(address(this));
+        if (wethBalance != 0) {
+            Constants.WETH.withdraw(wethBalance);
+        }
+    }
+
+    /// @notice Sets the maximum slippage vs 1:1 price accepted during withdraw.
+    /// @param _maximumSlippageBasisPoints the maximum slippage expressed in basis points (1/10_000)
+    function setMaximumSlippage(uint256 _maximumSlippageBasisPoints) external onlyGovernorOrAdmin {
+        require(_maximumSlippageBasisPoints <= Constants.BASIS_POINTS_GRANULARITY, "BalancerPCVDepositBase: Exceeds bp granularity.");
+        maximumSlippageBasisPoints = _maximumSlippageBasisPoints;
+        emit UpdateMaximumSlippage(_maximumSlippageBasisPoints);
+    }
+
+    /// @notice redeeem all assets from LP pool
+    function exitPool() external whenNotPaused onlyPCVController {
+        uint256 bptBalance = IWeightedPool(poolAddress).balanceOf(address(this));
+        if (bptBalance != 0) {
+            IVault.ExitPoolRequest memory request;
+
+            // Uses encoding for exact BPT IN withdrawal using all held BPT
+            bytes memory userData = abi.encode(IWeightedPool.ExitKind.EXACT_BPT_IN_FOR_TOKENS_OUT, bptBalance);
+            request.assets = poolAssets;
+            request.minAmountsOut = new uint256[](poolAssets.length); // 0 minimums
+            request.userData = userData;
+            request.toInternalBalance = false; // use external balances to be able to transfer out tokenReceived
+
+            vault.exitPool(poolId, address(this), payable(address(this)), request);
+
+            _burnFeiHeld();
+
+            emit ExitPool();
+        }
+    }
+
+    /// @notice claim BAL rewards associated to this PCV Deposit.
+    /// Note that if dual incentives are active, this will only claim BAL rewards.
```

Comment:
Is it possible to handle dual incentives case?

Reply:
we can pass an array of distributionIds, amounts, merkle proofs, token addresses, and token distributors, but at this point we might as well call the MerkleOrchard directly 😅 I'm not very convinced about this claimRewards function btw, there are so many parameters, it's going to be more convenient to make an offline documentation on how to claim (or make a front-end / use Balancer's front-end after they allow to claim for another address)